add device id and provisioning scripts

jayteemo opened the pull request:
add script to use provisioning API.
add script to automate claiming and provisioning.
IRIS-6795

jayteemo committed:
add device id and provisioning scripts

add script to use provisioning API.
add script to automate claiming and provisioning.
IRIS-6795

jayteemo commented:
Example usage:
`python3 claim_and_provision_device.py --api_key $API_KEY --stage dev --ca=./ca.pem --ca_key=ca_prv_key.pem --verbose`

provide your api key and the PEM files for _your_ CA cert and it's private key.
the script requires that the device is running the nrf provisioning sample... you can provide a hex file and it will flash the device.

plskeggs commented on python/modem-firmware-1.3+/claim_and_provision_device.py:
```diff
@@ -0,0 +1,522 @@
+plain = False
+verbose = False
+serial_timeout = 1
+aws_ca  = "-----BEGIN CERTIFICATE-----\nMIIDQTCCAimgAwIBAgITBmyfz5m/jAo54vB4ikPmljZbyjANBgkqhkiG9w0BAQsF\nADA5MQswCQYDVQQGEwJVUzEPMA0GA1UEChMGQW1hem9uMRkwFwYDVQQDExBBbWF6\nb24gUm9vdCBDQSAxMB4XDTE1MDUyNjAwMDAwMFoXDTM4MDExNzAwMDAwMFowOTEL\nMAkGA1UEBhMCVVMxDzANBgNVBAoTBkFtYXpvbjEZMBcGA1UEAxMQQW1hem9uIFJv\nb3QgQ0EgMTCCASIwDQYJKoZIhvcNAQEBBQADggEPADCCAQoCggEBALJ4gHHKeNXj\nca9HgFB0fW7Y14h29Jlo91ghYPl0h ...
```
You need to rebase and switch to using `ca_certs.py` I think.

plskeggs commented on python/modem-firmware-1.3+/claim_and_provision_device.py:
```diff
@@ -0,0 +1,522 @@
+    # claim device
+    print(local_style('Claiming device...'))
+    api_res = nrf_cloud_diap.claim_device(args.api_key, attest_tok)
+    nrf_cloud_diap.print_api_result("Claim device response", api_res, args.verbose)
```
If the device is already claimed, what happens?

Should we check for an error and abort here?

plskeggs commented on python/modem-firmware-1.3+/claim_and_provision_device.py:
```diff
@@ -0,0 +1,522 @@
+    print(hivis_style('\nProvisioning API URL: ' + nrf_cloud_diap.set_dev_stage(args.stage)))
+
+    # TODO - unclaim first?
+    #api_res = nrf_cloud_diap.unclaim_device(args.api_key, dev_uuid)
```
Maybe add an argument to the CLI to unclaim first.

plskeggs commented on python/modem-firmware-1.3+/claim_and_provision_device.py:
```diff
@@ -0,0 +1,522 @@
+    # write CA cert(s) to modem
+    if args.install_ca or args.coap or args.prov:
+        print(error_style('\n*** Installing CA certs currently not supported with provisioning sample ***\n'))
+        if 0: # TODO: how to write certs via provisioning sample?
```
How about, if this is chosen on the cli, first flash the at client, install the CA certs, then flash the provisioning client?

jayteemo replied:
IRIS-6918

plskeggs commented on python/modem-firmware-1.3+/claim_and_provision_device.py:
```diff
@@ -0,0 +1,522 @@
+    if not prov_id:
+        error_exit('Failed to obtain provisioning cmd ID')
+
+    # TODO: create provisioning command to install AWS root CA?
```
yes, and CoAP if requested

jayteemo replied:
i wonder if it would just be better to update `nrf_provisioning` to always add both CAs?
my update last week of the `ca_certs.py` was not quite correct... 
the **AWS CA** is used for _provisioning over http_ and the coap CA is used for _provisioning over coap_ (which i didn't realize was implemented).

currently, `nrf_provisioning` writes the cert for the configured protocol:
```
#if defined(CONFIG_NRF_PROVISIONING_HTTP)
		/* Certificate for nRF HTTP Provisioning */
		const uint8_t *cert = cert_amazon_root_ca1_pem;
		const size_t cert_size = cert_amazon_root_ca1_pem_size;
#else
		/* Certificate for nRF COAP Provisioning */
		const uint8_t *cert = cert_coap_root_ca_pem;
		const size_t cert_size = cert_coap_root_ca_pem_size;
#endif
```

created IRIS-6918

plskeggs replied:
Yeah, I have thought about that too (just always doing both).

jayteemo replied:
ok, i will leave it up to you to coordinate such a change with the `nrf_provisioning` code owners 😺

plskeggs commented on python/modem-firmware-1.3+/nrf_cloud_diap.py:
```diff
@@ -0,0 +1,120 @@
+#!/usr/bin/env python3
+#
+# Copyright (c) 2021 Nordic Semiconductor ASA
```
2023

plskeggs commented on python/modem-firmware-1.3+/nrf_cloud_provision.py:
```diff
@@ -66,33 +76,56 @@ def parse_args():
     return requests.put(req, json=json_payload, headers=hdr)
 
+def provision_device(api_key, dev_id, sub_type, tags, fw_types, cert_pem_str):
+    hdr = {'Authorization': 'Bearer ' + api_key,
```
Seems like this could allow a non-Nordic person to provision a device to dev or beta.  Maybe not a new problem, but perhaps a security risk?

jayteemo replied:
if they have an API key for dev or beta, they can do whatever they want :)

plskeggs commented on python/modem-firmware-1.3+/claim_and_provision_device.py:
```diff
@@ -0,0 +1,522 @@
+    global ser
+    ser.close()
+
+def get_attestation_token():
```
Maybe a future improvement, but, it would be good to be able to do this without USB too.  For example, the customer provides a csv file containing one or more attestation tokens.  The script would assume the device somewhere out there in the world is on and already has the provisioning client (or a customer application that includes the provisioning library) installed as well as the proper provisioning root CA.  The script would just interact with the provisioning REST API.

jayteemo replied:
IRIS-6919

jayteemo committed:
CA cert cleanup

There are only 2 CAs required: the AWS CA and the nRF Cloud CoAP CA.
Device Identity and Provisioning uses the AWS CA for the REST connection and the CoAP CA for provisioning over CoAP.

jayteemo commented:
added a commit to cleanup the CA cert usage.
removed the `--prov` option from the installer script since it is not needed.

plskeggs commented on python/modem-firmware-1.3+/nrf_cloud_provision.py:
```diff
@@ -31,7 +31,17 @@ class ProvisionResult(Enum):
 from modem_credentials_parser import write_file
 
-API_URL = "https://api.nrfcloud.com/v1/"
+DEV_STAGE_DICT = {'dev':     '.dev.',
```
I just occurred to me that if we are going to add dev and beta support here, we might as well add ca certs for them too.  See my branch: https://github.com/plskeggs/utils/tree/feature-coap-all-stages

jayteemo replied:
they use a different CA for dev and beta?
what is the reason for that?

jayteemo replied:
IRIS-6923

plskeggs replied:
I did not ask.

SoorajRajasekharan commented on python/modem-firmware-1.3+/claim_and_provision_device.py:
```diff
@@ -0,0 +1,526 @@
+        if 0: # TODO: how to write certs via provisioning sample?
+            print(local_style('Installing CA cert(s)...'))
+            if args.coap:
+                modem_ca = ca_certs.nrf_cloud_coap_ca + ca_certs.aws_ca
```
Can it install the CoAP CA certificate based on the stage? CoAP CA certificates are different for these three environments.

jayteemo replied:
CA cert installation has been implemented.
You will need to build your provisioning client FW with the following options:
```
CONFIG_NRF_PROVISIONING_RX_BUF_SZ=2048
CONFIG_SHELL_BACKEND_SERIAL_RX_RING_BUFFER_SIZE=2048
CONFIG_SHELL_CMD_BUFF_SIZE=2048
```

plskeggs replied:
Should these Kconfig changes be added officially to the provisioning client prj.conf?

jayteemo replied:
not my decision.
you could ask Juha, as he helped me get it working: https://nordicsemi.atlassian.net/browse/NCSDK-23596

plskeggs replied:
OK, I was unaware these were not your changes but came from Juha.

jayteemo replied:
the changes to these python scripts are from me based on comments from Juha in the linked ticket.

SoorajRajasekharan commented on python/modem-firmware-1.3+/claim_and_provision_device.py:
```diff
@@ -0,0 +1,526 @@
+    parser.add_argument("--ca_key", type=str,
+                        help="Filepath to your CA's private key PEM",
+                        default="./ca_prv_key.pem")
+    parser.add_argument("--csv", type=str,
```
provision.csv file was not generated after secure provisioning.

jayteemo replied:
i forgot to remove this parameter, it was carried over from the device credentials installer script.

jayteemo replied:
removed

SoorajRajasekharan commented on python/modem-firmware-1.3+/claim_and_provision_device.py:
```diff
@@ -0,0 +1,526 @@
+    if not prov_id:
+        error_exit('Failed to obtain provisioning cmd ID')
+
+    # TODO: create provisioning command to install AWS root CA?
```
After the "Client Certificate Command," the next step should be to include the "Finish command" in the script in order to change the device status to "provisioned" in nrfcloud and restart the device after provisioning. Finish command should be added as the last command to change the status from Awaiting to Provisioned.
![image](https://github.com/nRFCloud/utils/assets/125363902/2c134e7f-fbdd-4a8a-a2cc-26c5e4224f8c)

jayteemo replied:
ah i see, "finish" in the test plan.  i didn't realize that was a provisioning command 😆

jayteemo replied:
<img width="677" alt="image" src="https://github.com/nRFCloud/utils/assets/4138391/04bb78f7-dba3-40c4-b6e2-e87f65e64c9a">
Updated

jayteemo committed:
Add provisioning finished cmd

jayteemo committed:
add CA cert installation

jayteemo commented:
please re-review/test with the new commits.

jayteemo committed:
add claim and prov script to readme
